Add variable disabled attribute and interactions with it.

marisaleung opened the pull request:
### Resolves

Implements the spec discussed here: https://github.com/LLK/scratch-blocks/issues/975
Deleted variables reappear on toolbox refresh if they are deleted and still exist in the workspace.

### Proposed Changes

- Add a disabled attribute to VariableModel.
- When a variable is disabled, it does not show up in dropdowns or the toolbox. If a dropdown in the workspace already has that variable selected, that instance will remain.
- A variable is disabled when the user tries to delete a variable at the same time that there is a workspace instance of it.
- When the variable is enabled, it will reappear in the dropdowns and toolbox.
- A variable is enabled when the toolbox refreshes and when it is selected in a dropdown on the workspace.
- A variable is deleted if it is disabled and the last instance of it in the workspace is removed. For example:
    1. If the last instance is a block and it is deleted.
    2. If the last instance is a dropdown with that variable selected, and the a different variable is selected. 
- Undo and redo features are supported.

### Reason for Changes

Made to match the spec in the issue above. This is so variable can come back in the toolbox if a user accidentally deleted it when that variable is still used in the workspace.

marisaleung committed:
Variables reappear on refreshes where variables aren't deleted.

rachel-fenichel commented:
>A variable is enabled when the toolbox refreshes and it is selected in a dropdown.

Nit: selected in a dropdown on the workspace, right?

> A variable is deleted if it is disabled and the last instance of it in the workspace is removed. For example:

What about variables that exist across multiple workspaces?  If the last instance of the variable is deleted in the currently selected workspace but it's still in use in another workspace?

rachel-fenichel commented on core/block.js:
```diff
@@ -278,6 +280,12 @@ Blockly.Block.prototype.dispose = function(healStack) {
     }
   } finally {
     Blockly.Events.enable();
+    if (variableField) {
```
Why does this happen after the event enable?

marisaleung replied:
It is after the enable for two reasons:
1. I want the VAR_DELETE event to be fired if it happens.
2. deleteDisabledVariableIfUnused() checks if there are any workspace uses of the variable, after enable, this block is gone so the number of uses will be 0. If it is put before, the the disable() the number of uses will include this block and be 1.

rachel-fenichel commented on core/events.js:
```diff
@@ -1086,6 +1095,43 @@ Blockly.Events.VarRename.prototype.run = function(forward) {
 };
 
 /**
+ * Class for a variable disable event.
```
Add a note that this is scratch-specific/that mainline Blockly has no "disabled variable" concept.

marisaleung replied:
Done.

rachel-fenichel commented on core/events.js:
```diff
@@ -1086,6 +1095,43 @@ Blockly.Events.VarRename.prototype.run = function(forward) {
+ */
+Blockly.Events.VarDisable.prototype.run = function(forward) {
+  var workspace = this.getEventWorkspace_();
+  if (!workspace) return;
```
Put the return on a new line and add braces around it.

The style guide technically says this is okay, but it's not the style that the rest of Blockly uses.

marisaleung replied:
Thanks! Done.

marisaleung commented:
Correct, updated comment. And this is a good point. We decided to ignore the multiple workspaces for now and implement everything else since the multiple workspaces requires a listener for the vm. If Scratch decides they like the variable reappearing on refresh, then we will continue and add that listener.

rachel-fenichel commented on core/variable_model.js:
```diff
@@ -87,6 +94,21 @@ Blockly.VariableModel.prototype.getId = function() {
+ * Enable the variable.
+ */
+Blockly.VariableModel.prototype.enable = function() {
+  this.disabled = false;
```
No event is fired when you enable a variable?

marisaleung replied:
This is a good point. I found a bug from that thank you! Added a an enable event.

rachel-fenichel commented on core/field_variable.js:
```diff
@@ -163,9 +163,11 @@ Blockly.FieldVariable.dropdownCreate = function() {
   var options = [];
-  for (var i = 0; i < variableModelList.length; i++) {
+  for (var i = 0, variable; variable = variableModelList[i]; i++) {
+    if (!variable.disabled || this.getValue() == variable.name) {
```
You check ` if(!variable.disabled)` quite a few times in this PR.  

It would be easier to parse `if (variable.isEnabled())`.  Also (per offline discussion) switch to storing `enabled` instead of `disabled` to avoid double negatives.

marisaleung replied:
Done

rachel-fenichel commented on core/workspace.js:
```diff
@@ -401,19 +388,35 @@ Blockly.Workspace.prototype.deleteVariableById = function(id) {
+    this.deleteVariableInternal_(variable);
+  }
+};
+
```
Remove extra newlines.

marisaleung replied:
Done

rachel-fenichel commented on core/workspace.js:
```diff
@@ -401,19 +388,35 @@ Blockly.Workspace.prototype.deleteVariableById = function(id) {
-    uses[i].dispose(true, false);
+  // If the variable is disabled, the delete occurs with another event such as
+  // removing the last workspace instance of the variable.
+  if (!variable.disabled) Blockly.Events.setGroup(true);
```
Add braces and move the inside of the if to a new line, here and below.

marisaleung replied:
Done.

marisaleung commented:
Uploading more changes:
- Added a var_enable event
- Renamed 'disabled' attribute to 'enabled' to avoid double negatives.
- Added a Blockly.Variables.allowEnable attribute so when flyoutCategory is run right after a variable is disabled, it doesn't automatically re-enable the variable.
    - Also added "Blockly.Variables.allowEnable = false" in Blockly.Events.VarCreate() because if a VarDisable() occurred in the same group, the following would happen during an undo:
        1. Disable variable
        2. Blockly.Variables.allowEnable = false
        3. Refresh toolbox -> run flyoutCategory
        4. Blockly.Variables.allowEnable = true
        5. Delete Variable
        6. Refresh toolbox -> run flyoutCategory
        7. Enable variable that was just disabled.

paulkaplan commented:
Is it possible to pass the `enabled` flag in through XML properties in the `<variables>` section?

marisaleung commented:
Since this is a change specifically for scratch-blocks feel free to add one, but we will not be adding it to blockly. Scratch gui should work without it in the variables section.

paulkaplan commented:
I'm seeing an issue when I build this into GUI:
1. Create a variable
2. Immediately delete the variable (without dragging any blocks out of the palette)
3. Switch to the backdrop, see the variable comes back but says "undefined"
![image](https://user-images.githubusercontent.com/654102/28370044-4f65c5f0-6c67-11e7-9757-d101384e6864.png)

Can't tell where this is problem is occurring, I'll need to track it down deeper.

rschamp commented:
This PR will be useful when we do implement this, but probably a bit too out of date to actually use as-is.
